README.md
# video_controls

A wrapper around `video_player ` to show customizable controls for Flutter videos.

Customization options are being implemented, but the package is in a working state. 

### Features

- Pause/Play button
- Full screen (only tested on Web)
- Slider for seeking and line to show the buffer
- Seeking without lag
- Mute button
- Controls that fade out when the mouse is still or not on the video
  - On mobile, tapping the video shows the controls for a few seconds

### Screenshots

| ![](https://github.com/Levi-Lesches/video_controls/blob/main/doc/demo4.gif) | ![](https://github.com/Levi-Lesches/video_controls/blob/main/doc/demo1.gif) |
| ------------------------------------------------------------ | ------------------------------------------------------------ |



### Customization options 

Coming soon! But you can use regular ThemeData properties for now, since this package uses standard widgets. 

### Usage

```dart
import "package:flutter/material.dart";

import "package:video_controls/video_controls.dart";

const String url = "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4";

class HomePage extends StatefulWidget {
  @override
  HomePageState createState() => HomePageState();
}

class HomePageState extends State<HomePage> {
  final VideoController controller = VideoController.network(url);

  @override
  Widget build(BuildContext context) => Scaffold(
    appBar: AppBar(title: Text("video_controls Demo")),
    body: Center(
      child: ConstrainedBox(
        constraints: const BoxConstraints(maxWidth: 750),
        child: ListView(
          padding: const EdgeInsets.all(10),
          children: [
            Text(
              "My Video",
              textAlign: TextAlign.center,
              style: Theme.of(context).textTheme.headline3
            ),
            const SizedBox(height: 30),
            VideoPlayer(controller),
          ]
        )
      )
    )
  );

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }
}

void main() => runApp(MaterialApp(home: HomePage()));

```





CHANGELOG.md
## 0.1.0 - 2021-07-30
### Added
- Controls overlay fades out when not focusing on the video
- Controls appear when tapping on the screen (for mobile)
- Controls fade out when the cursor doesn't move
- Added README
- Improved example

### Fixed
- Controls overlay is transparent
- Fixed bug when exiting full screen with system back button (or ESC)

## 0.0.2 - 2021-07-22
## 0.0.1 - 2021-07-22

lib/src/player.dart
import "dart:async";

import "package:flutter/material.dart";
import "package:video_player/video_player.dart" as plugin;

import "controller.dart";
import "controls.dart";

/// A video player with controls overlayed. 
/// 
/// This widget is meant to replace the VideoPlayer from `package:video_player`.
/// This widget puts a [VideoControls] widget on top, and manages full-screen
/// transitions. However, the logic still depends on VideoPlayerController,
/// but with an added property. Use [VideoController] from this package instead.
class VideoPlayer extends StatefulWidget {
	/// The controller backing this player. 
	final VideoController controller;

	/// Creates a video player with controls overlayed. 
	const VideoPlayer(this.controller);

	@override
	VideoPlayerState createState() => VideoPlayerState();
}

/// The state for a [VideoPlayer]. 
/// 
/// Manages full-screen transitions and overlays the controls atop the video.
class VideoPlayerState extends State<VideoPlayer> {
	/// The [VideoController] for this video. 
	VideoController get controller => widget.controller;

	/// Whether this video is ready to be played. 
	bool get isReady => controller.value.isInitialized;

	/// Whether the mouse is hovering over the video.
	bool isHovering = false;

	/// Fades the controls out when this timer is run. 
	/// 
	/// This is used to add a delay before fading out. Animations wouldn't work
	/// because the animation would start too soon. With a timer, we can cancel
	/// it and restart it whenever we want. 
	Timer? disableControls;

	@override
	void initState() {
		super.initState();
		init();
	}

	@override
	void dispose() {
		disableControls?.cancel();
		super.dispose();
	}

	/// Initializes the video, if it isn't already, and rebuilds. 
	Future<void> init() async {
		if (isReady) return;
		await controller.initialize();
		setState(() {});
	}

	@override
	Widget build(BuildContext context) => AspectRatio(
		aspectRatio: controller.value.aspectRatio,
		child: Listener(
			behavior: HitTestBehavior.translucent,
			onPointerDown: (_) => disableControls?.cancel(),
			onPointerUp: (_) => disableControls = Timer(
				const Duration(seconds: 3), 
				() => setState(() => isHovering = false)
			),
			onPointerHover: (_) {
				setState(() => isHovering = true);
				disableControls?.cancel();
				disableControls = Timer(
					const Duration(seconds: 2), 
					() => setState(() => isHovering = false)
				);
			},
			child: Stack(
				children: [
					// Video background
					if (isReady) plugin.VideoPlayer(controller)
					else Container(color: Colors.grey.shade200.withOpacity(0.1)),

					// Controls foreground
					AnimatedSwitcher(
						duration: const Duration(milliseconds: 500),
						child: controller.value.isPlaying && !isHovering ? Container() : Align(
						alignment: Alignment.bottomCenter,
						child: !isReady ? const LinearProgressIndicator() : SizedBox(
							height: 100,
							child: VideoControls(
								controller,
								openFullScreen: openFullScreen,
								closeFullScreen: closeFullScreen,
							)
						) 
					)
					)
				]
			)
		)
	);

	/// Closes the full screen player
	Future<bool> closeFullScreen() async {
		Navigator.of(context).pop(controller.value); 
		return false;
	}

	/// Opens the video in full screen.
	/// 
	/// For some reason, we need to re-initialize the video when the full screen
	/// dialog is dismissed. So we save the state obtained by [showDialog] and 
	/// re-initialize [controller] to match.
	Future<void> openFullScreen() async {
		controller.isFullScreen = true;
		final plugin.VideoPlayerValue video = await showDialog(
			context: context,
			builder: (_) => WillPopScope(
				onWillPop: closeFullScreen,
				child: Scaffold(body: Center(child: VideoPlayer(controller)))
			)
		);
		controller.isFullScreen = false;
		// BUG: We shouldn't have to re-initialize every time!
		await controller.initialize();
		await controller.seekTo(video.position);
		if (video.isPlaying) await controller.play();
		setState(() {});
	}
}


lib/src/controls.dart
import "package:flutter/material.dart";
import "package:video_player/video_player.dart";
import "controller.dart";

extension on Duration {
  String get timestamp => "${inMinutes.toString().padLeft(2, '0')}:"
    "${(inSeconds % 60).toString().padLeft(2, '0')}";
}

/// A widget that shows controls for a video. 
/// 
/// This widget provides interactivity for videos backed by 
/// `package:video_player`. It should be overlaid on top of [VideoPlayer], and 
/// is backed by a [VideoController], which must be initialized before 
/// this widget is created. 
class VideoControls extends StatefulWidget {
  /// The `video_player` plugin's [VideoController].
  final VideoController controller;

  /// The callbacks used to open and close full screen mode. 
  /// 
  /// Because full-screen mode depends on the Widget tree, that is left to the 
  /// widget on top of this one. 
  final VoidCallback openFullScreen, closeFullScreen;

  /// Creates a widget to show video controls.
  /// 
  /// [controller] must be initialized before creating this widget 
  const VideoControls(
    this.controller, 
    {
      required this.openFullScreen,
      required this.closeFullScreen,
    }
  );

  @override
  VideoControlsState createState() => VideoControlsState();
}

/// The state for the video controls. 
/// 
/// This state uses [controller] from `package:video_player` to read and control
/// the state of the video, which can be accessed through [video].
class VideoControlsState extends State<VideoControls> {
  /// The `video_player` plugin's [VideoController].
  VideoController get controller => widget.controller;

  /// The current state of the video.
  VideoPlayerValue get video => controller.value;

  /// The position the user is trying to seek to. 
  /// 
  /// This allows the UI to move the slider while still playing the video, and 
  /// only actually loading the selected position when the user releases. 
  /// 
  /// Null when the user is not moving the slider. 
  double? seekValue;

  /// The current position of the buffer. 
  Duration get bufferedPosition => video.buffered.isEmpty
    ? Duration.zero : video.buffered.last.end;

  /// Whether the video is currently muted. 
  bool get isMuted => video.volume == 0;

  /// The timestamp in `current` / `end` form. 
  /// 
  /// If the user is moving the slider (when [seekValue] is non-null), the
  /// `current` time shows the position of the slider. 
  String get timestamp {
    final Duration currentPosition = seekValue == null 
      ? video.position : doubleToDuration(seekValue!);
    return "${currentPosition.timestamp} / ${video.duration.timestamp}";
  }

  /// Updates the value of [video] whenever the controller updates.
  void listener() => setState(() {});

  @override
  void initState() {
    super.initState();
    controller.addListener(listener);
  }

  @override
  void dispose() {
    widget.controller.removeListener(listener);
    super.dispose();
  }

  /// Converts a [Duration] object to a double. 
  /// 
  /// Used to allow a [Duration] object to be used with a [Slider].
  double durationToDouble(Duration duration) => 
    duration.inSeconds.toDouble();

  /// Converts a double to a [Duration] object.
  /// 
  /// This allows a [Slider] to output [Duration] values. 
  Duration doubleToDuration(double position) => 
    Duration(minutes: position ~/ 60, seconds: (position % 60).truncate());

  @override
  Widget build(BuildContext context) => IconTheme(
    data: const IconThemeData(color: Colors.white),
    child: Container(
      color: const Color.fromRGBO(0, 0, 0, 0.5),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          Stack(
            children: [
              SliderTheme(
                data: SliderThemeData(thumbShape: SliderComponentShape.noThumb),
                child: Slider(
                  max: durationToDouble(video.duration),
                  value: durationToDouble(bufferedPosition),   
                  onChanged: null,         
                )
              ),
              Slider(
                max: durationToDouble(video.duration),
                value: seekValue ?? durationToDouble(video.position),
                onChanged: (double value) => setState(() => seekValue = value),
                onChangeEnd: (double value) {
                  setState(() => seekValue = null);
                  controller.seekTo(doubleToDuration(value));
                },
              ),
            ]
          ),
          Row(
            children: [
              IconButton(
                icon: Icon(video.isPlaying ? Icons.pause : Icons.play_arrow ),
                onPressed: video.isPlaying ? controller.pause : controller.play,
              ),
              IconButton(
                icon: Icon(isMuted ? Icons.volume_off : Icons.volume_up),
                onPressed: () => widget.controller.setVolume(isMuted ? 1 : 0),
              ),
              Text(
                timestamp,
                style: Theme.of(context).textTheme.caption
                  ?.copyWith(color: Colors.white),
              ),
              const Spacer(),
              IconButton(
                icon: Icon(
                  controller.isFullScreen 
                    ? Icons.fullscreen_exit : Icons.fullscreen
                ),
                onPressed: controller.isFullScreen
                  ? widget.closeFullScreen: widget.openFullScreen,
              )
            ],
          )
        ]
      )
    )
  );
}


example/main.dart
import "package:flutter/material.dart";

import "package:video_controls/video_controls.dart";

const String url = "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4";

class HomePage extends StatefulWidget {
  @override
  HomePageState createState() => HomePageState();
}

class HomePageState extends State<HomePage> {
  final VideoController controller = VideoController.network(url);

  @override
  Widget build(BuildContext context) => Scaffold(
    appBar: AppBar(title: const Text("video_controls Demo")),
    body: Center(
      child: ConstrainedBox(
        constraints: const BoxConstraints(maxWidth: 750),
        child: ListView(
          padding: const EdgeInsets.all(10),
          children: [
            Text(
              "My Video",
              textAlign: TextAlign.center,
              style: Theme.of(context).textTheme.headline3
            ),
            const SizedBox(height: 30),
            VideoPlayer(controller),
          ]
        )
      )
    )
  );

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }
}

void main() => runApp(MaterialApp(home: HomePage()));
